### ISSUE TITLE
React: `_jsx is not defined` error

### ISSUE THREAD

**Issue by shilman** [storybookjs/storybook#12952]
_Originally posted by @nandosangenetto in https://github.com/storybookjs/storybook/issues/12881#issuecomment-718969472_

I'm getting the same error that @DiFuks is getting, `_jsx is not defined`.

![image](https://user-images.githubusercontent.com/7468379/97622265-c3132280-1a02-11eb-8ddb-5158a4ca6ff2.png)

I've created a small repo where you can reproduce this bug: https://github.com/nandosangenetto/storybook-bug-repro

I've tried to remove all addons, as @DiFuks did, but it didn't work.

What is odd is that it works when I run `npm run storybook` (`start-storybook -p 6006`), but fails when I try to build it with `npm run build-storybook` (`build-storybook`).

**Comment by shilman** [comment]
@nandosangenetto i think this is unrelated to #12881, or only tangentially related. i copied `.babelrc` into the `.storybook` directory and the error went away. LMK if that works for you.

i did, however, see something very troubling, which is:
- i tried modifying the `main.js` config to remove the addons (which you also tried)
- i rebuilt `storybook-static` and the addons were still being installed
- i tried messing around including deleting `node_modules/.cache` and rebuilding `storybook-static`

in the end deleting `node_modules`, reinstalling, and rebuilding worked. i've never seen this before, and i didn't have time to follow up. is there anything special about your project that might cause this?

**Comment by nandosangenetto** [comment]
I cannot recall anything special. I'll copy the `.babelrc` file into the `.storybook` directory to see if it builds successfully in our CircleCI step. I'm going to let you know if it works.

Let me ask you, what node and npm version are you using?

**Comment by nandosangenetto** [comment]
Although it's building, I cannot switch between stories/templates. Is it working for you?

I recorded a video showing it, the only way I could change the component was by refreshing the page, since it was changing the URL but not the canvas.

**Comment by shilman** [comment]
Try building with --no-dll

**Comment by nandosangenetto** [comment]
It worked! 🎉  Thank you @shilman 

Just to recap, the solution was to copy the `.babelrc` file into `.storybook` folder, and build storybook with the `--no-dll` option.

**Comment by DiFuks** [comment]
@shilman Yes, It worked. But I don't understand why it worked. My .babelrc
```json
{
  "presets": ["next/babel"],
  "env": {
    "development": {
      "plugins": ["@babel/plugin-transform-react-display-name"]
    }
  }
}
```

**Comment by DiFuks** [comment]
I removed .babelrc from .storybook and changed main.js.  It worked:

```js
const { TsconfigPathsPlugin } = require('tsconfig-paths-webpack-plugin');

module.exports = {
  stories: ['../src/components/**/*.stories.tsx'],
  addons: ['@storybook/addon-docs'],
  babel: async (options) => ({
    ...options,
    plugins: [
      ...options.plugins,
      '@babel/plugin-transform-react-jsx'
    ]
  }),
  webpackFinal: (config) => {
    config.resolve.plugins.push(new TsconfigPathsPlugin());

    return config;
  }
};
```
@shilman @nandosangenetto  Can you try please?

**Comment by shilman** [comment]
@DiFuks great point. perhaps there's a bug where dev mode picks up the project babelrc and production does not? if so, that's pretty bad! can you confirm?

**Comment by DiFuks** [comment]
@shilman I replaced the content of .babelrc in root path with `{}` and remove babel config from .storybook/main.js
And... Production build worked! It's very funny 😅

I posted a project that has the discussed issue: https://github.com/DiFuks/react-next-nest-boilerplate

**master** branch include this fix: https://github.com/storybookjs/storybook/issues/12952#issuecomment-719871776
**storybook** branch include empty .babelrc file in project root and empty babel config in .storybook/main.js
**storybook-bug** branch include original code with error

**Comment by Carduelis** [comment]
@shilman  @DiFuks 
I had the same issue (`_jsx is not defined`), I don't use `--no-dll` flag (but it's still yelling about `(node:8752) DeprecationWarning: DLL-related CLI flags are deprecated, see: %url%`.

Copying `<root>/.babelrc` to `<root>/.storybook/.babelrc` did help. I use `next`, btw. 
```
// .babelrc
{
  "presets": ["next/babel"],
  "plugins": ["react-require"]
}
```

Storybook version is `@storybook/react v6.1.0-alpha.35`

**Comment by DiFuks** [comment]
@Carduelis This solution didn't help? https://github.com/storybookjs/storybook/issues/12952#issuecomment-719871776

**Comment by shilman** [comment]
@Carduelis Please upgrade to the latest beta

**Comment by sibelius** [comment]
fixed for us on v6.1.0-rc.0

**Comment by DiFuks** [comment]
@sibelius Unfortunately, the problem has not been resolved.

**Comment by eric-burel** [comment]
Can reproduce on Vulcan Next https://github.com/VulcanJS/vulcan-next/issues/78

**Comment by Spy-Seth** [comment]
After upgrading to v6.0.1 my storybook production build got broken with the same error.

I use the proposed fix of @DiFuks and it fix my build

**Comment by shilman** [comment]
Try upgrading to 6.1?

```
npx sb upgrade
```

**Comment by ifree92** [comment]
```
"@storybook/addon-actions": "^6.1.5",
"@storybook/addon-essentials": "^6.1.5",
"@storybook/addon-links": "^6.1.5",
"@storybook/react": "^6.1.5",
```

I have the same issue.

The solution for me - just to copy the `.babelrc` file into `.storybook` folder.
E.g.: `cp .babelrc .storybook/`

BTW, I have tried to upgrade a storybook version via `npx sb upgrade`, but, actually, helpless.

**Comment by mariuslundgard** [comment]
I resorted to patching the babel config like so (in `main.js`):

```js
  ...
  babel: (options) => {
    options.presets = options.presets.map((preset) => {
      if (Array.isArray(preset) && preset[0].includes('@babel/preset-react')) {
        return [require.resolve('@babel/preset-react'), {runtime: 'classic'}]
      }

      return preset
    })

    return options
  },
  ...
```

**Comment by shilman** [comment]
Does anybody have a repro they can share with `6.1.6`? I'm unable to repro the above on my machine.

Here's what I think is going on:
- In our initialization code we try to import the JSX transform and use `classic` if that fails
- In our webpack config we use alias to force a single version of `react` using `require.resolve` (this changed recently https://github.com/storybookjs/storybook/pull/13195)

I think the logic of those two is somehow inconsistent. With a simple repro it should hopefully be easy to fix.

**Comment by Fi1osof** [comment]
> Does anybody have a repro they can share with `6.1.6`? I'm unable to repro the above on my machine.

@shilman here: https://github.com/prisma-cms/component-boilerplate/tree/76d2516757491068c3cfc1df5dc143ef69437b73
For repro error please remove .storybook/.babelrc

**Comment by shilman** [comment]
@Fi1osof thanks for this. When I do `npx sb upgrade` to get all the storybook packages to their latest version, it appears to work properly even without the `.storybook/.babelrc`. Can you verify on your machine?

**Comment by ifree92** [comment]
@shilman I have tried to upgrade via `npx sb upgrade` (and removed `.storybook/.babelrc` file), so I have faced with the same _jsx reference error.

The better solution is just to copy .babelrc into .storybook folder and that might work for now..

**Comment by Fi1osof** [comment]
@shilman i try `npx sb upgrade`, but still have error. But i have not this error here: https://github.com/prisma-cms/nextjs 
It's similar projects with minimal diffs. I think something with tsconfig.ts maybe. In https://github.com/prisma-cms/component-boilerplate/ modified.

**Comment by Spy-Seth** [comment]
@shilman I still have the error with `v6.1.6`

**Comment by rhatfield-bushel** [comment]
I'm experiencing this with `v6.1.6` and I found this issue while sorting through an [issue of my own](https://github.com/storybookjs/storybook/discussions/13278).

I used @DiFuks's example, and also needed a workaround from Chakra's own repo.
I was able to remove the duplicate `.babelrc` from the `.storybook` directory, and I no longer need to use `--no-dll`.

Below is my `main.js` file, and this is the minimum I found required for me to get storybook working without the `_jsx` issue. 

```javascript
const path = require('path')

const toPath = (_path) => path.join(process.cwd(), _path)

module.exports = {
  stories: ['../src/**/*.stories.mdx', '../src/**/*.stories.@(js|jsx|ts|tsx)'],
  addons: [
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    'storybook-addon-playroom',
    '@storybook/addon-a11y',
    '@storybook/addon-storysource',
    'storybook-addon-designs',
    '@storybook/addon-actions',
  ],

  // ! Taken from Storybook issue workaround
  // ! See: https://github.com/storybookjs/storybook/issues/12952#issuecomment-719871776
  babel: async (options) => ({
    ...options,
    plugins: [...options.plugins, '@babel/plugin-transform-react-jsx'],
  }),

  // ! Taken from Chakra's Storybook + Emotion 11 workaround
  // ! See: https://github.com/chakra-ui/chakra-ui/pull/2466/files#diff-cafe2123a72c4ce3a9f7e9ee4b0e188256eb02d7ec1e54fa2cadfac99b22f92b
  webpackFinal: async (config) => {
    return {
      ...config,
      resolve: {
        ...config.resolve,
        alias: {
          ...config.resolve.alias,
          '@emotion/core': toPath('node_modules/@emotion/react'),
          'emotion-theming': toPath('node_modules/@emotion/react'),
        },
      },
    }
  },
}

```

**Comment by shilman** [comment]
@Fi1osof I'm seeing a slightly different error which is `_jsxs is not defined`. Can you confirm? Also which version of `npm` are you using?

**Comment by Fi1osof** [comment]
> @Fi1osof I'm seeing a slightly different error which is `_jsxs is not defined`. Can you confirm? Also which version of `npm` are you using?

npm@6.14.8
But i using yarn. Try npx build-storybook too.

**Comment by Fi1osof** [comment]
> I'm experiencing this with `v6.1.6` and I found this issue while sorting through an [issue of my own](https://github.com/storybookjs/storybook/discussions/13278).
> 
> I used @DiFuks's example, and also needed a workaround from Chakra's own repo.
> I was able to remove the duplicate `.babelrc` from the `.storybook` directory, and I no longer need to use `--no-dll`.
> 
> Below is my `main.js` file, and this is the minimum I found required for me to get storybook working without the `_jsx` issue.
> 
> ```js
> 
> module.exports = {
>   // ...
>   // ! Taken from Storybook issue workaround
>   // ! See: https://github.com/storybookjs/storybook/issues/12952#issuecomment-719871776
>   babel: async (options) => ({
>     ...options,
>     plugins: [...options.plugins, '@babel/plugin-transform-react-jsx'],
>   }),
> // ...
> }
> ```

@rhatfield-bushel this helps me. And @babel/plugin-transform-react-jsx already exists in SB.
```
$ yarn why @babel/plugin-transform-react-jsx
yarn why v1.22.5
[1/4] Why do we have the module "@babel/plugin-transform-react-jsx"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@babel/plugin-transform-react-jsx@7.12.7"
info Reasons this module exists
   - "@storybook#addon-docs" depends on it
   - Hoisted from "@storybook#addon-docs#@babel#plugin-transform-react-jsx"
   - Hoisted from "@storybook#react#@babel#preset-react#@babel#plugin-transform-react-jsx"
info Disk size without dependencies: "60KB"
```
I do not know why this not works nativly.

**Comment by radihuq** [comment]
I'm in the process of refactoring our component library and I had this piece of commented out code from the old component:

```
// ...
// OLD COMPONENT
// /** @jsx jsx */
// import { jsx } from '@emotion/core';

// import React from 'react';
// // import { useTheme } from 'emotion-theming';
// ...
```

We used to [need that first line](https://emotion.sh/docs/typescript#emotionreact) but since we no longer use EmotionJS, removing that line resolved the error for me.

Silly mistake

**Comment by pwfisher** [comment]
Adding this minimal `/.storybook/.babelrc` worked for me (using next.js):
```
// .babelrc
{
  "presets": ["next/babel"]
}
```

**Comment by Yago** [comment]
> Adding this minimal `/.storybook/.babelrc` worked for me (using next.js):
> 
> ```
> // .babelrc
> {
>   "presets": ["next/babel"]
> }
> ```

Same solution for me, but with a less duplicated configuration:

```
// ./.storybook/.babelrc

{
  "extends": "../.babelrc"
}

```

**Comment by ndelangen** [comment]
This is fixed in storybook 7.0 beta by not providing a default babel config.
